stub and nf-test created for raxmlng

bhargava-morampalli opened the pull request:
<!--
# nf-core/modules pull request

Many thanks for contributing to nf-core/modules!

Please fill in the appropriate checklist below (delete whatever is not relevant).
These are the most common things requested on pull requests (PRs).

Remember that PRs should be made against the master branch.

Learn more about contributing: [CONTRIBUTING.md](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
-->

## PR checklist

Closes #7676

- [x] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
- [ ] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [ ] If necessary, include test data in your PR.
- [x] Remove all TODO statements.
- [ ] Emit the `versions.yml` file.
- [ ] Follow the naming conventions.
- [ ] Follow the parameters requirements.
- [ ] Follow the input/output options guidelines.
- [ ] Add a resource `label`
- [ ] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
  - For modules:
    - [x] `nf-core modules test <MODULE> --profile docker`
    - [ ] `nf-core modules test <MODULE> --profile singularity`
    - [ ] `nf-core modules test <MODULE> --profile conda`
  - For subworkflows:
    - [ ] `nf-core subworkflows test <SUBWORKFLOW> --profile docker`
    - [ ] `nf-core subworkflows test <SUBWORKFLOW> --profile singularity`
    - [ ] `nf-core subworkflows test <SUBWORKFLOW> --profile conda`

SPPearce reviewed (changes requested):
I know you are just swapping to nf-test, but think this module could do with some improvements.

I think this module should have a meta map. Then we can use meta.id as the default prefix, as normal.
Is the `--model` a required argument? As then it should be an input channel; please don't define extra `task.ext.foo` values.
Can you make the optional file in the stub (bonus points if you make it conditional on something being passed to args, if that is how it decides to make it.

SPPearce commented:
If you are not comfortable adjusting the module, I can do that for you.

SPPearce commented:
Oh, and don't forget to remove the pytest files and the entry from the pytest config file.

SPPearce commented:
And while you are there, you can remove the pytest files and entry for `roary`.

bhargava-morampalli commented:
@SPPearce 
I have done the changes you suggested. Can you please take a look at the changes when you have a chance?
1) I wanted to modify the module myself - so made changes as you suggested
2) I deleted the entries in pytest config file and deleted the pytest files as well

SPPearce commented on modules/nf-core/raxmlng/main.nf:
```diff
@@ -7,28 +7,55 @@ process RAXMLNG {
 
     script:
-    def args = task.ext.args ?: ''
+    def args = task.ext.args ?: params.raxmlng_args ?: ''
```
```suggestion
    def args = task.ext.args ?: ''
```

SPPearce commented on modules/nf-core/raxmlng/main.nf:
```diff
@@ -7,28 +7,55 @@ process RAXMLNG {
     script:
-    def args = task.ext.args ?: ''
+    def args = task.ext.args ?: params.raxmlng_args ?: ''
+    def prefix = task.ext.prefix ?: meta.id
```
```suggestion
    def prefix = task.ext.prefix ?: "${meta.id}"
```

SPPearce commented on modules/nf-core/raxmlng/main.nf:
```diff
@@ -7,28 +7,55 @@ process RAXMLNG {
-    def args = task.ext.args ?: ''
+    def args = task.ext.args ?: params.raxmlng_args ?: ''
+    def prefix = task.ext.prefix ?: meta.id
+    if (!meta.id) {
+        error "Input meta map does not contain 'id'. Received: ${meta}"
+    }
```
```suggestion
```
We assume that the meta.id is present (the only part of the meta that we assume exists)

SPPearce commented on modules/nf-core/raxmlng/main.nf:
```diff
@@ -7,28 +7,55 @@ process RAXMLNG {
+        error "Input meta map does not contain 'id'. Received: ${meta}"
+    }
+    // Use a dedicated param for stub testing the bootstrap output scenario
+    def bootstrapping = args.contains('--bootstrap') || args.contains('--bs-trees') || params.test_stub_bootstrap == true
+    def touch_support = bootstrapping ? "touch ${prefix}.raxml.support" : ""
```
```suggestion
    def touch_support = args.contains('--bootstrap') || args.contains('--bs-trees') ? "touch ${prefix}.raxml.support" : ""
```

SPPearce commented on modules/nf-core/raxmlng/meta.yml:
```diff
@@ -13,24 +13,41 @@ tools:
       - "*.raxml.bestTree":
-          type: file
-          description: A phylogeny in Newick format
+          type: map
+          description: Groovy Map containing sample information
```
This should still be the previous description and type:

bhargava-morampalli replied:
resolved - made changes as suggested

SPPearce commented on modules/nf-core/raxmlng/meta.yml:
```diff
@@ -13,24 +13,41 @@ tools:
       - "*.raxml.support":
-          type: file
-          description: A phylogeny in Newick format with bootstrap values
+          type: map
```
This should still be the previous description and type

bhargava-morampalli replied:
resolved - made changes as suggested

SPPearce commented on modules/nf-core/raxmlng/tests/main.nf.test:
```diff
@@ -0,0 +1,93 @@
+            assertAll (
+                { assert process.success },
+                // This will show phylogeny and versions, but not phylogeny_bootstrapped
+                { assert snapshot(process.out).match("stub_no_bootstrap") }
```
```suggestion
                { assert snapshot(process.out).match() }
```

SPPearce reviewed (changes requested):
You can use a modules_args parameter to pass the bootstrapping option to the `args`, to test that directly rather than adding a new extra parameter.
See https://github.com/nf-core/modules/blob/c9c3ef86c1892413b3c86fb38c4e39fd7288512f/modules/nf-core/ataqv/ataqv/tests/main.nf.test for instance.

SPPearce commented on modules/nf-core/raxmlng/main.nf:
```diff
@@ -7,28 +7,51 @@ process RAXMLNG {
     """
+
+    stub:
+    def args = task.ext.args ?: params.raxmlng_args ?: ''
```
```suggestion
    def args = task.ext.args ?: ''
```

SPPearce commented on modules/nf-core/raxmlng/main.nf:
```diff
@@ -7,28 +7,51 @@ process RAXMLNG {
+    stub:
+    def args = task.ext.args ?: params.raxmlng_args ?: ''
+    def prefix = task.ext.prefix ?: meta.id
+    if (!meta.id) {
+        error "Input meta map does not contain 'id'. Received: ${meta}"
+    }
+    // Use a dedicated param for stub testing the bootstrap output scenario
```
```suggestion
```

SPPearce commented on modules/nf-core/raxmlng/main.nf:
```diff
@@ -7,28 +7,51 @@ process RAXMLNG {
+
+    stub:
+    def args = task.ext.args ?: params.raxmlng_args ?: ''
+    def prefix = task.ext.prefix ?: meta.id
```
```suggestion
    def prefix = task.ext.prefix ?: "${meta.id}"
```

SPPearce commented:
@bhargava-morampalli, are you able to get back to this?

bhargava-morampalli commented on modules/nf-core/raxmlng/main.nf:
```diff
@@ -7,28 +7,55 @@ process RAXMLNG {
-    path "*.raxml.bestTree", emit: phylogeny
-    path "*.raxml.support" , optional:true, emit: phylogeny_bootstrapped
-    path "versions.yml"    , emit: versions
+    tuple val(meta), path("*.raxml.bestTree")              , emit: phylogeny
+    tuple val(meta), path("*.raxml.support"), optional:true, emit: phylogeny_bootstrapped
+    path "versions.yml"                                    , emit: versions
```
resolved - made changes as suggested

bhargava-morampalli commented:
> @bhargava-morampalli, are you able to get back to this?

@SPPearce I cannot see any more changes requested, can you please take a look when you have time and let me know if there are more changes required?

SPPearce commented on modules/nf-core/raxmlng/main.nf:
```diff
@@ -7,28 +7,47 @@ process RAXMLNG {
-    path "*.raxml.bestTree", emit: phylogeny
-    path "*.raxml.support" , optional:true, emit: phylogeny_bootstrapped
-    path "versions.yml"    , emit: versions
+    tuple val(meta), path("*.raxml.bestTree")              , emit: phylogeny
+    tuple val(meta), path("*.raxml.support")               , emit: phylogeny_bootstrapped, optional:true
+    path "versions.yml"                                    , emit: versions
```
```suggestion
    tuple val(meta), path("*.raxml.bestTree"), emit: phylogeny
    tuple val(meta), path("*.raxml.support") , emit: phylogeny_bootstrapped, optional:true
    path "versions.yml"                      , emit: versions
```

SPPearce commented on modules/nf-core/raxmlng/main.nf:
```diff
@@ -7,28 +7,47 @@ process RAXMLNG {
     """
+
+    stub:
+    def args = task.ext.args ?: ''
```
```suggestion
```

SPPearce commented on modules/nf-core/raxmlng/tests/main.nf.test:
```diff
@@ -0,0 +1,91 @@
+        }
+
+        then {
+            assertAll (
+                { assert process.success },
+                // Check phylogeny output (tuple: meta, path)
+                { assert process.out.phylogeny.size() == 1 },
+                { assert process.out.phylogeny[0][0].id == 'test' }, // Check meta.id passed through
+                { assert process.out.phylogeny[0][1].toString().endsWith(".raxml.bestTree") },
+                { assert process.out.phylogeny[0][1].toString().contains("test") }, // Check if prefix 'test' was used
+                // Check optional bootstrapped output (should be empty)
+                { assert process.out.phylogeny_bootstrapped.isEmpty() },
+                // Check versions
+                { assert process.out.versions.size() == 1 },
+                { assert process.out.versions[0].toString().endsWith("versions.yml") },
+                { assert snapshot(process.out.versions[0]).match("versions_yml") } // Snapshot specific file
+            )
```
Can this not just be snapshotting the entirety of process.out?

bhargava-morampalli replied:
@SPPearce I tried using process.out to capture everything but I am getting md5 mismatch errors in snapshots. that is why I resorted to checking individual files instead
I tried random seed as well but still get md5 mismatches
If I definitely need to use process.out, I am out of my depth here. I have tried everything I can. Is it possible for you to take a look or someone else in nf-core to resolve this?

bhargava-morampalli replied:
@SPPearce 
I looked into some similar issues and this seems like a known issue. 
1) They suggest to either use only this in assertions
`{ assert process.success }`
2) ignore them or use specific assertions to files only as I have done
What do you think? I want to learn the best practice in these scenarios

This is the error I am getting.
`java.lang.RuntimeException: Different Snapshot:
expected                                                                                            observed
       3         "0": [                                                                                    3         "0": [
       4             [                                                                                     4             [
       5                 {                                                                                 5                 {
       6                     "id": "test"                                                                  6                     "id": "test"
       7                 },                                                                                7                 },
       8                 "test.raxml.bestTree:md5,618e1dba0db6ca6b0815ffce6c6939cd"                        8                 "test.raxml.bestTree:md5,be5ceeca8d9040631791126437051b06"
       9             ]                                                                                     9             ]
      10         ],                                                                                       10         ],
      11         "1": [                                                                                   11         "1": [
      12                                                                                                  12
      13         ],                                                                                       13         ],
---                                                                                                 ---
      17         "phylogeny": [                                                                           17         "phylogeny": [
      18             [                                                                                    18             [
      19                 {                                                                                19                 {
      20                     "id": "test"                                                                 20                     "id": "test"
      21                 },                                                                               21                 },
      22                 "test.raxml.bestTree:md5,618e1dba0db6ca6b0815ffce6c6939cd"                       22                 "test.raxml.bestTree:md5,be5ceeca8d9040631791126437051b06"
      23             ]                                                                                    23             ]
      24         ],                                                                                       24         ],
      25         "phylogeny_bootstrapped": [                                                              25         "phylogeny_bootstrapped": [
      26                                                                                                  26
      27         ],                                                                                       27         ],`

SPPearce replied:
Hi @bhargava-morampalli ,
Basically yes, you do need to use different assertions at that point. Generally I just check the contents or the name of the file if it is binary. 
I added a default random seed (using the code from umitools/dedup), so that if not specified it will use 42.
Also the nextflow.config was missing to make the modules_args actually work, so the bootstrapping test wasn't actually working.
Take a look at the files that I just pushed to your branch.

SPPearce reviewed (commented):
Sorry, been on vacation for the last two weeks.

LouisLeNezet commented on modules/nf-core/raxmlng/main.nf:
```diff
@@ -3,32 +3,52 @@ process RAXMLNG {
+    def touch_support = args.contains('--bootstrap') || args.contains('--bs-trees') ? "touch ${prefix}.raxml.bootstraps" : ""
+    """
+    # Create stub output files
+    touch ${prefix}.raxml.bestTree
```
If it's optional is there a way to know when `.bestTree` files are generated from the arguments ?

SPPearce replied:
Hmm, I think that it is either bestTree or the bootstrap file (from the other tests), so could possibly make the stub either/or. But I'm not certain.

SPPearce replied:
I've swapped to make the file dependent on the bootstrap or not.

LouisLeNezet commented on modules/nf-core/raxmlng/tests/main.nf.test.snap:
```diff
@@ -0,0 +1,139 @@
+                    ]
+                ],
+                "phylogeny_bootstrapped": [
+                    
```
Why is it empty ?

SPPearce replied:
Excellent catch, because args wasn't in the stub.

LouisLeNezet commented on tests/modules/nf-core/roary/main.nf:
Why did you delete this file.
Was the migration already done ?

LouisLeNezet reviewed (changes requested):
A few comments before approving.

LouisLeNezet reviewed (approved):
Good to go for me !
